controllers/hpa/hpa_reconciler.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2021 The KServe Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hpa
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	"github.com/kserve/kserve/pkg/constants"
+	"github.com/kserve/kserve/pkg/utils"
+	v2beta2 "k8s.io/api/autoscaling/v2beta2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("HPAReconciler")
+
+// HPAReconciler is the struct of Raw K8S Object
+type HPAReconciler struct {
+	client client.Client
+	scheme *runtime.Scheme
+	HPA    *v2beta2.HorizontalPodAutoscaler
+}
+
+func NewHPAReconciler(client client.Client,
+	scheme *runtime.Scheme, runtimeMeta metav1.ObjectMeta, mmDeploymentName string, mmNamespace string) *HPAReconciler {
+	return &HPAReconciler{
+		client: client,
+		scheme: scheme,
+		HPA:    createHPA(runtimeMeta, mmDeploymentName, mmNamespace),
+	}
+}
+
+func getHPAMetrics(metadata metav1.ObjectMeta) []v2beta2.MetricSpec {
+	var metrics []v2beta2.MetricSpec
+	var utilization int32 = constants.DefaultCPUUtilization
+
+	annotations := metadata.Annotations
+	resourceName := corev1.ResourceCPU
+
+	if value, ok := annotations[constants.TargetUtilizationPercentage]; ok {
+		utilizationInt, _ := strconv.Atoi(value)
+		utilization = int32(utilizationInt)
+	}
+
+	if value, ok := annotations[constants.AutoscalerMetrics]; ok {
+		resourceName = corev1.ResourceName(value)
+	}
+
+	metricTarget := v2beta2.MetricTarget{
+		Type:               "Utilization",
+		AverageUtilization: &utilization,
+	}
+
+	ms := v2beta2.MetricSpec{
+		Type: v2beta2.ResourceMetricSourceType,
+		Resource: &v2beta2.ResourceMetricSource{
+			Name:   resourceName,
+			Target: metricTarget,
+		},
+	}
+
+	metrics = append(metrics, ms)
+	return metrics
+}
+
+func createHPA(runtimeMeta metav1.ObjectMeta, mmDeploymentName string, mmNamespace string) *v2beta2.HorizontalPodAutoscaler {
+	var minReplicas int32
+	minReplicas = int32(constants.DefaultMinReplicas)
+	maxReplicas := int32(constants.DefaultMinReplicas)
```

Comment:
Should this be `DefaultMaxReplicas`?

Reply:
The default maximum number of replicas is the same as the DefaultMin number of replicas, but if you have a better direction, I'd always love to hear from you.

Reply:
Ignore my comment then, I simply thought it was a typo :) I'm very new with this codebase so I still need to adapt to this coding style (given that `maxReplicas := int32(constants.DefaultMinReplicas)` is an implicit contract on `constants.DefaultMinReplicas`, I thought there was already a specific constant for the max number of replicas).